### ISSUE TITLE
AwsGenerateStsCredentials/AwsGenerateCredentials

### ISSUE THREAD

**Issue by MushiTheMoshi** [hashicorp/vault-client-go#158]
# Expected Behavior
```.go
	awsMountPath := "aws_v1" <<< I need this to be v1 or v2
	resp2, err := client.Secrets.AwsGenerateStsCredentials(
		context.Background(),
		name,
		awsMountPath,
		vault.WithToken(resp.Auth.ClientToken), <<< collected before
	)
```
client.Secrets.AwsGenerateStsCredentials return aws credentials

Followed this example:https://github.com/hashicorp/vault-client-go/blob/main/docs/SecretsApi.md#awsgeneratestscredentials
Looks like **awsMountPath** have changed to a type **vault.RequestOption** , can you please provide an example how to use it?

# Current Behavior
erroring

What is the current behavior?
Type is not as mentioned in the documentation

# Failure Information
cannot use awsMountPath (variable of type string) as type vault.RequestOption in argument to client.Secrets.AwsGenerateStsCredentials

Please include the version of Vault binary and the version of `vault-client-go` you're using.
require github.com/hashicorp/vault-client-go v0.3.0

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

https://github.com/hashicorp/vault-client-go/blob/main/docs/SecretsApi.md#awsgeneratecredentials 
![image](https://user-images.githubusercontent.com/13651419/236328695-6ec1ef5b-00d8-4163-b4b3-d50bce0a6a1f.png)

# Additional Information

**Comment by averche** [comment]
Hi @MushiTheMoshi!

The generated docs are not quite correct here. Thanks for pointing it out, I will make sure to update them.

Please use `vault.WithMountPath(...)` to adjust the mount path. An example can be found [in the readme](https://github.com/hashicorp/vault-client-go#authentication):

```go
resp, err := client.Secrets.AwsGenerateStsCredentials(
	context.Background(),
	name,
	vault.WithToken(resp.Auth.ClientToken),   // this line can be skipped if the client is authenticated with SetToken(...) earlier
	vault.WithMountPath("aws_v1"),
)
```

**Comment by MushiTheMoshi** [comment]
thanks for the quick response!

**Comment by MushiTheMoshi** [comment]
hey

**Comment by MushiTheMoshi** [comment]
I am getting permission denied:
![image](https://user-images.githubusercontent.com/13651419/236513879-32233f2f-33bd-446c-8312-2a60853ed354.png)

whereas with the same mountpath in python I get credentials, how is that possible?

![image](https://user-images.githubusercontent.com/13651419/236514404-c986a67e-fbe6-44a2-beee-680e963dffe4.png)

**Comment by MushiTheMoshi** [comment]
I think it is getting authenticated since `resp.Auth.ClientToken` returns a vault token , also I have used the same to login using vault cli and token is working, can't share image now, but I will on Monday.

**Comment by MushiTheMoshi** [comment]
hi there thanks for all the help, please find below:
![image](https://user-images.githubusercontent.com/13651419/236878031-289c30dc-3839-41e6-8ae6-2c738a15942b.png)

**Comment by averche** [comment]
I think I know what the issue is. Could you please try the following steps.

1. Replace the library version in you `go.mod`:
```
github.com/hashicorp/vault-client-go v0.3.0
```
with
```
github.com/hashicorp/vault-client-go v0.2.1-0.20230427174350-ff6f4d893b42
```

2. Run `go mod tidy`

3. Replace your function call with the following:

```go
	resp2, err := client.Secrets.AwsGenerateCredentials2(
		context.Background(),
		name,
		schema.AwsGenerateCredentials2Request{
			RoleArn:         "< insert arn here >",
			RoleSessionName: "< insert session name >",
			Ttl:             "< insert ttl here >",
		},
		vault.WithMountPath("aws_v2"),
	)

```

For the definitions of RoleArn, RoleSessionName, and Ttl, please see the [documentation](https://developer.hashicorp.com/vault/api-docs/secret/aws#generate-credentials)

If these steps work, I will update the library accordingly.

**Comment by MushiTheMoshi** [comment]
Hi there, I have updated as per your suggestion , now I am getting a type error, please find below

![image](https://user-images.githubusercontent.com/13651419/236909841-9181c43c-cbf1-4c94-a218-67a7b0f23f49.png)


![image](https://user-images.githubusercontent.com/13651419/236909821-714dfc23-278a-465a-abea-ee159446fa4e.png)

**Comment by averche** [comment]
The screenshot you posted is missing a "2" at the end (should be `Secrets.AwsGenerateCredentials2` not `Secrets.AwsGenerateCredentials`.

**Comment by MushiTheMoshi** [comment]
yeah, I noticed late, I have updated to the correct, getting this:
![image](https://user-images.githubusercontent.com/13651419/236910922-3643017b-bbe2-4973-970e-bbf24e81e754.png)

![image](https://user-images.githubusercontent.com/13651419/236910960-aa45b92a-69ef-4d55-bf86-9dfbca505b50.png)

**Comment by averche** [comment]
It's missing a "Request" at the end (should be `schema.AwsGenerateCredentials2Request`)

**Comment by MushiTheMoshi** [comment]
I have updated accordingly, same issue from before, `permission denied`
![image](https://user-images.githubusercontent.com/13651419/236911612-ee1526a3-26f9-4668-bde8-f6b9b234ecb5.png)

ERROR:
![image](https://user-images.githubusercontent.com/13651419/236911649-075dda20-6314-4a5f-aeba-0a8ca7eca97c.png)

**Comment by averche** [comment]
Please try one more thing, add "Sts" in both the request and the request struct:

```go
	resp2, err := client.Secrets.AwsGenerateStsCredentials2(
		context.Background(),
		name,
		schema.AwsGenerateStsCredentials2Request{
			RoleArn:         "< insert arn here >",
			RoleSessionName: "< insert session name >",
			Ttl:             "< insert ttl here >",
		},
		vault.WithMountPath("aws_v2"),
	)
```

**Comment by averche** [comment]
At this point, my guess is that the token does not have permissions to generate STS credentials or the request is not structured correctly. You can see the policies associated with the token by running the following:

```sh
 vault token lookup | grep policies
```

To test whether generating STS credential works outside of the library, you can try the following in command line:

```sh
vault write aws_v2/sts/<insert_role_name> \
    role_arn="<insert_role_arn>" \
    role_session_name="<insert_role_session_name>" \
    ttl="3600s"
```

**Comment by MushiTheMoshi** [comment]
This is giving same error, then why when I use python script is working?

![image](https://user-images.githubusercontent.com/13651419/236927341-e24f7e55-2d1a-430b-bbfd-0efed6471660.png)


![image](https://user-images.githubusercontent.com/13651419/236927816-0ff6b014-32c6-40af-b9b2-fed0c5bd42d4.png)

**Comment by averche** [comment]
Your command line should have `aws_v2` instead of `aws`.

Looking at the python code, it appears you are using HVAC, specifically calling [this function](https://github.com/hvac/hvac/blob/a4a972af021eb1e646f12092f5e9bb9f8a0bccfc/hvac/api/secrets_engines/aws.py#L327-L407). Since you are not passing `"sts"` override, the python code is generating regular credentials (_not_ STS)!

The corresponding function to call in the Go library is:

```go
	resp2, err := client.Secrets.AwsGenerateCredentials(
		context.Background(),
		name,
		vault.WithMountPath("aws_v2"),
	)
```

The corresponding CLI command is:

```sh
vault read aws_v2/sts/<insert_role_name>
```

**Comment by MushiTheMoshi** [comment]
true, I can see result with `vault read aws_v2/sts/<insert_role_name>` though when using `client.Secrets.AwsGenerateCredentials` this is not working... I got permission denied

**Comment by MushiTheMoshi** [comment]
vault cli:
![image](https://user-images.githubusercontent.com/13651419/236936522-20d56634-fb2e-452b-845b-394fdfd16163.png)

go script:
![image](https://user-images.githubusercontent.com/13651419/236936761-084d4cb2-fea7-4d86-a5c9-c5c917933a03.png)

**Comment by averche** [comment]
To debug this further, please add the following in your request. This will print the request before it is sent out:
```go
	resp2, err := client.Secrets.AwsGenerateCredentials(
		context.Background(),
		name,
		vault.WithMountPath("aws_v2"),
		vault.WithRequestCallbacks(func(r *http.Request) {
			log.Println(*r)
		}),
	)
```

Then run:
```sh
vault read -output-curl-string aws_v2/creds/<insert_role_name>
```
... and compare the outputs. Ideally the two request paths should match.

**Comment by MushiTheMoshi** [comment]
hi, looks like endpoint - role is not honoring capitalization, my role contains capital letters that is the only difference...

**Comment by averche** [comment]
I've just tried this locally and I see that the case is not modified (capital letters are preserved):

```go
	_, err = client.Secrets.AwsGenerateCredentials(
		ctx,
		"Test-Role-123",
		vault.WithMountPath("aws_v2"),
		vault.WithRequestCallbacks(func(r *http.Request) {
			log.Println(*r)
		}),
	)
```

```
GET http://127.0.0.1:8200/v1/aws_v2/creds/Test-Role-123
```

Possibly the role you are passing in is not correct?

**Comment by MushiTheMoshi** [comment]
you are right, this `resp.Auth.Policies` is changing capitalization, sorry about it, thanks again for all the help and hints to troubleshoot this issue, will close the ticket and again thank you so much for your help and quick response...
